### PR TITLE
[RL] Remove the RL-specific batch configuration

### PR DESCRIPTION
**Pull request by acisseJZhong**
Description:
  - Remove RL BatchConfig and use trainer batch-shape fields directly.
  - Derive local_batch_size from num_tokens_per_microbatch_per_dp_rank /
    max_context_length.
  - Preserve the existing row packing and round-robin DP distribution
    behavior.
  - Validate that the token budget contains a whole number of rows.
  - Keep RL optimizer-step boundaries controlled by
    num_prompts_per_train_step.

  Testing:
  - 232 RL tests passed, 2 skipped.

### CODE REVIEW

File: torchtitan/experiments/rl/controller.py [hunk 1]
```diff
@@ -804,6 +803,7 @@ async def run(self) -> None:
 
         # batcher
         batcher = async_loop.batcher.build(
+            training=self.config.trainer.training,
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/controller.py]: Can we just pass the (two) fields needed by batcher, instead of passing in all training config? O/w it's unnecessary information leakage.

File: torchtitan/experiments/rl/examples/alphabet_sort/config_registry.py [hunk 2]
```diff
@@ -500,7 +497,11 @@ def rl_grpo_qwen3_14b() -> Controller.Config:
                 warmup_steps=2,
                 decay_type="linear",
             ),
-            training=TrainingConfig(dtype="bfloat16"),
+            training=TrainingConfig(
+                num_tokens_per_microbatch_per_dp_rank=2 * 2048,
+                max_context_length=2048,
+                dtype="bfloat16",
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/examples/alphabet_sort/config_registry.py:503]: unrelated: let's remove any dtype=bf16 in TrainingConfig

File: torchtitan/experiments/rl/components/batcher.py [hunk 3]
```diff
@@ -89,12 +73,23 @@ def __init__(
...
         *,
+        num_tokens_per_microbatch_per_dp_rank: int,
+        max_context_length: int,
         num_prompts_per_train_step: int,
         dp_degree: int,
         pad_id: int,
     ) -> None:
-        self.local_batch_size = config.batch.local_batch_size
-        self.seq_len = config.batch.seq_len
+        self.seq_len = max_context_length
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/components/batcher.py:82]: seq_len used to mean "num_tokens_per_microbatch_per_dp_rank"?? should rename anyway

File: torchtitan/experiments/rl/components/batcher.py [hunk 4]
```diff
@@ -89,12 +73,23 @@ def __init__(
...
+        self.local_batch_size, remainder = divmod(
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/components/batcher.py:83]: remove `local_batch_size`

**Review by tianyu-l — CHANGES_REQUESTED** [review]

File: torchtitan/experiments/rl/components/batcher.py (hunk 4 above)
Comment (acisseJZhong) [reply torchtitan/experiments/rl/components/batcher.py:83]: I intentionally kept self.seq_len and self.local_batch_size to honor the packing algorithm we had in RL. See https://github.com/pytorch/torchtitan/blob/b953a3fd79fa3309a2475f409b34c42cdf6f0690/torchtitan/experiments/rl/components/batcher.py#L289

We can improve our packing algorithm later in a separate PR. @felipemello1

File: torchtitan/experiments/rl/components/batcher.py (hunk 3 above)
Comment (acisseJZhong) [reply torchtitan/experiments/rl/components/batcher.py:82]: before `self.seq_len = config.batch.seq_len`, so this change should align with what we had before?

**Review by tianyu-l — APPROVED** [review]
@wwwjn when updating https://github.com/pytorch/torchtitan/pull/3938, please remove batch size / seq len separation, now that the trainer accepts folded dim.

**Comment by wwwjn** [comment]
> @wwwjn when updating #3938, please remove batch size / seq len separation, now that the trainer accepts folded dim.

Thanks, let me follow up and update #3938

**Comment by acisseJZhong** [comment]
RL integration tests was skipped, not sure why. Ran it locally and it passed.
